core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -53,11 +53,17 @@
 class EntryHashSet<K, V> extends EntryArray<K, V> {
 
     /*-------------- Constants --------------*/
-    static final int INVALID_KEY_HASH = 0; // because memory is initially zeroed
+
     static final int DEFAULT_COLLISION_CHAIN_LENGTH = 3;
     // HASH - the key hash of this entry (long includes the update counter)
     private static final int HASH_FIELD_OFFSET = 2;
-    private static final int KEY_HASH_BITS = 32; // Hash needs to be an integer
+    private static final int KEY_HASH_BITS = 33; // Hash needs to be an integer + INVALID_KEY_HASH
+
+    // key hash may have any positive integer value including zero. However, initially
+    // all array's memory is zeroed, including key hashes and their update counters.
+    // The following INVALID_KEY_HASH is relevant only when the update counter is non-zero.
+    static final int INVALID_KEY_HASH = (1 << (KEY_HASH_BITS - 1));
```

Comment:
All hash values are supposed to be valid. We can take one bit from the counter that will represent the validity of the hash.
Zero will be invalid so that the default entry value will always be invalid.

Reply:
I now see that `KEY_HASH_BITS` is 33 which means that the invalid bit is outside of the hash "zone".
(1) This calculation creates an integer overflow.  Change it to:
```suggestion
    static final long INVALID_KEY_HASH = (1L << (KEY_HASH_BITS - 1));
```
(2) Why not use the `UnionCodec` for this?
```java
private final UnionCodec hashCodec = new UnionCodec(
        Integer.SIZE, // for the hash value
        1, // invalid hash flag
        UnionCodec.INVALID_BIT_SIZE // counter use whatever bits that are left
); 
```

Reply:
I saw it before the review and fixed that